src/java.base/share/classes/com/sun/crypto/provider/ConstructKeys.java

```diff
@@ -161,29 +158,42 @@ private static final PrivateKey constructPrivateKey(byte[] encodedKey,
      * @return a secret key constructed from the encodedKey.
      */
     private static final SecretKey constructSecretKey(byte[] encodedKey,
-                                              String encodedKeyAlgorithm)
-    {
-        return (new SecretKeySpec(encodedKey, encodedKeyAlgorithm));
+            int ofs, int len, String encodedKeyAlgorithm) {
+        return (new SecretKeySpec(encodedKey, ofs, len, encodedKeyAlgorithm));
     }
 
     static final Key constructKey(byte[] encoding, String keyAlgorithm,
-                                  int keyType)
-        throws InvalidKeyException, NoSuchAlgorithmException {
-        Key result = null;
+            int keyType) throws InvalidKeyException, NoSuchAlgorithmException {
         switch (keyType) {
         case Cipher.SECRET_KEY:
-            result = ConstructKeys.constructSecretKey(encoding,
-                                                      keyAlgorithm);
-            break;
+            return ConstructKeys.constructSecretKey(encoding, 0,
+                    encoding.length, keyAlgorithm);
         case Cipher.PRIVATE_KEY:
-            result = ConstructKeys.constructPrivateKey(encoding,
-                                                       keyAlgorithm);
-            break;
+            return ConstructKeys.constructPrivateKey(encoding, keyAlgorithm);
         case Cipher.PUBLIC_KEY:
-            result = ConstructKeys.constructPublicKey(encoding,
-                                                      keyAlgorithm);
-            break;
+            return ConstructKeys.constructPublicKey(encoding, keyAlgorithm);
+        default:
+            throw new RuntimeException("Unsupported key type");
+        }
+    }
+
+    static final Key constructKey(byte[] encoding, int ofs, int len,
+            String keyAlgorithm, int keyType)
+            throws InvalidKeyException, NoSuchAlgorithmException {
+        if (keyType == Cipher.SECRET_KEY) {
+            return ConstructKeys.constructSecretKey(encoding, ofs, len,
+                    keyAlgorithm);
+        } else {
+            if (ofs == 0 && len == encoding.length) {
+                return constructKey(encoding, keyAlgorithm, keyType);
+            } else {
+                byte[] encoding2 = Arrays.copyOfRange(encoding, ofs, ofs + len);
```

Comment:
Right, maybe there is no constructor which takes offset and length due to trying to avoid the cost of validating them. I recall seeing other *Spec classes whose list of constructors are similar to X509EncodedKeySpec/PKCS8EncodedKeySpec as well.